File: PAT甲级真题练习/1039. Course List for Student-超时.cpp
/*
 @Date    : 2018-01-27 19:17:10
 @Author  : 酸饺子 (changzheng300@foxmail.com)
 @Link    : https://github.com/SourDumplings
 @Version : $Id$
*/

/*
https://www.patest.cn/contests/pat-a-practise/1039
 */

#include <iostream>
#include <cstdio>
#include <string>
#include <vector>
#include <map>
#include <set>

using namespace std;

int main(int argc, char const *argv[])
{
    int N, K;
    scanf("%d %d", &N, &K);
    int Ni, c;
    map<string, set<int>> data;
    string name;
    for (int i = 0; i != K; ++i)
    {
        scanf("%d %d", &c, &Ni);
        for (int j = 0; j != Ni; ++j)
        {
            cin >> name;
            data[name].insert(c);
        }
    }
    vector<string> query;
    for (int i = 0; i != N; ++i)
    {
        cin >> name;
        query.push_back(name);
    }

    for (auto name : query)
    {
        cout << name << " " << data[name].size();
        for (auto it = data[name].begin(); it != data[name].end(); ++it)
            printf(" %d", *it);
        putchar('\n');
    }
    return 0;
}

File: PAT甲级真题练习/1039. Course List for Student-超时.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2017-12-06 16:05:13
# @Author  : 酸饺子 (changzheng300@foxmail.com)
# @Link    : https://github.com/SourDumplings
# @Version : $Id$

'''
https://www.patest.cn/contests/pat-a-practise/1039
'''


def main():
    temp = input().split(' ')
    N = int(temp[0])
    K = int(temp[1])

    students = {}

    for i in range(K):
        c_info = input().split(' ')
        c = int(c_info[0])
        n = int(c_info[1])
        if n:
            namelist = input().split(' ')
            for each_name in namelist:
                if each_name in students.keys():
                    students[each_name].append(c)
                else:
                    students[each_name] = [c]

    if N:
        query = input().split(' ')
        for each_name in query:
            if each_name not in students.keys():
                print("%s 0" % each_name)
                continue
            print("%s %d" % (each_name, len(students[each_name])), end='')
            students[each_name].sort()
            for each_course in students[each_name]:
                print(" %d" % each_course, end='')
            print()


if __name__ == '__main__':
    main()
